libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/SyncManager.java

```diff
@@ -437,23 +459,59 @@ private void syncDownSosl(SyncState sync, SyncUpdateCallback callback) throws Ex
 		}
 		return arr;
 	}
+
+    private Set<String> toSet(JSONArray jsonArray) throws JSONException {
+        Set<String> set = new HashSet<String>();
+        for (int i=0; i<jsonArray.length(); i++) {
+            set.add(jsonArray.getJSONArray(i).getString(0));
+        }
+        return set;
+    }
 	
-	private void saveRecordsToSmartStore(String soupName, JSONArray records)
+	private void saveRecordsToSmartStore(String soupName, JSONArray records, MergeMode mergeMode)
 			throws JSONException {
-		// Save to SmartStore
+        // Gather ids of dirty records
+        Set<String> idsToSkip = null;
```

Comment:
Approach: we get the ids of ALL the dirty records and skip them when saving records fetched from the server.
We don't do a:
select id where dirty and id in (ids fetched from server) 
because the list of ids could be to long and we would have to chunk up the query etc. 

We don't expect to many records changed locally, so the full list should be "small" most of the time.